### PR TITLE
416 user api

### PR DESCRIPTION
**Pull request by jum-s**
this PR

- create a user model, reusing db api
- create a user db independent from things
- implement couchdb views for users
- implement a findOrCreate user model method as a promise
- create routes for userInfo access (by _id)

### Questions in the air/ To discuss

- one user document per provider seems the most simple way to treat a user on the platform, unifying every providers under a single user doc would introduce unnecessary complexity
- shall user db be public access aka do user [routes](https://github.com/TransforMap/data.transformap.co/compare/master...jum-s:416-user-api?expand=1#diff-638c9659c027fffede2dfb2d2a0e5821R11) need some sort of authentication?
- [data validation](œhttps://github.com/TransforMap/data.transformap.co/compare/master...jum-s:416-user-api?expand=1#diff-638c9659c027fffede2dfb2d2a0e5821R11) seems light, but since only external providers post to the database it looks enough to me, is it to you?

### CODE REVIEW

File: test/user-model.js [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+  })
+  describe('findOrCreateUser', function () {
+    username = userNewDoc.contact.name
+
+    it('should be a function that returns an object', function (done) {
+      User.findOrCreateUser.should.be.a.Function()
+      User.findOrCreateUser(userNewDoc).should.be.an.Object()
+      done()
+    })
+    it('should return an object with the passed type', function (done) {
```
Comment (almereyda) [inline test/user-model.js:59]: I have cloned your branch locally and ran the tests. This one fails:

```

  31 passing (788ms)                                                                                                                    
  1 failing                                                                                          

  1) User model findOrCreateUser should return an object with the passed type: 
     TypeError: Cannot read property 'should' of undefined                     
      at test/user-model.js:62:20                                              
      at tryCatcher (node_modules/nano-blue/node_modules/bluebird/js/main/util.js:26:23)                                                                       
      at Promise._settlePromiseFromHandler (node_modules/nano-lue/node_modules/bluebird/js/main/promise.js:510:31)                                            
      at Promise._settlePromiseAt (node_modules/nano-blue/node_modules/bluebird/js/main/promise.js:584:18)                                                     
      at Async._drainQueue (node_modules/nano-blue/node_modules/bluebird/js/main/async.js:128:12)                                                              
      at Async._drainQueues (node_modules/nano-blue/node_modules/bluebird/js/main/async.js:133:10)                                                             
      at Immediate.Async.drainQueues (node_modules/nano-blue/node_modules/bluebird/js/main/async.js:15:14)                                                     

```

**Review by almereyda — CHANGES_REQUESTED** [review]
I am not against merging, but as the test that is failing has been introduced with this PR, it seems favourable to see this PR working completely. Then #48 can be updated and have the mockups removed for final review of the OAuth interfaces. Or is the tested method `findOrCreateUser` unused?

Additionally I could imagine a lightweight documentation of the features introduced by this PR and how to use them.

To answer the questions, which we already scratched in the last sitin:

1. If we ever intended to build a more encompassing user profiles system, these records would be a good way to collect service links and authorisations. For the moment we just need to keep the link between username and address, nothing else.
2. [User records](https://hack.allmende.io/s/transformaps-20170914-development#users-records) will only be exposed as names, not as email addresses.
3. Validation could be extended, once we expect further responses. Yet I would even argue to remove the email regex and accept any string here, as this validation will most likely have happened at the upstream identity provider already.

File: server/controllers/users/routes.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+const _ = __.require('lib', 'utils')
+const User = require('./lib/users')
+const error_ = __.require('lib', 'error')
+
+module.exports = {
+  generateRoutes: function () {
+    return {
+      "users": {
+        // TODO implement ensureAuthenticated from branch 425 oauth
+        post: function (req, res) {
```
Comment (acorbi) [inline server/controllers/users/routes.js:12]: @jum-s @almereyda As per our discussion from yesterday and as currently implemented on https://github.com/TransforMap/transformap-editor/pull/42/files#diff-f8eba5e5189b7491d79351405c4f5d73R684, the user endpoint should be extended with an `update` method (via PUT) so the user's property `agreedTos` can be altered once the user agrees with TOS

Comment (almereyda) [reply server/controllers/users/routes.js:12]: Agree. Similar to https://github.com/TransforMap/data.transformap.co/blob/master/server/controllers/things/things.js#L39

**Comment by jum-s** [comment]
So far, User record are immutable because it only stores ID provider references. It strictly implements OCP, and KISS

Now, if contact information on the provider server are updated, how does data.tranformap.io keep user store updated, since the user store db has no versioning? [User store](https://github.com/TransforMap/data.transformap.co/pull/49/files#diff-5562bacf9d68b79104b40a5abfa30ae7R8) provides a provider URL. To be a URI you could append the username, and gets another way to cantact the user
An update endpoint would open a whole new architecture (with all CRUD options).
I do not quite understand the necessity of this `agreedTos` storage while it could be implicit, with a  'less data storing' solution.
When a new user signs up, they are sent to the provider page agreement, and findOrCreateUser as already implemented
Then, if a new user is created, they are redirected to TOS with an 'Agree' button.
If the user clicks on agree, then it would continue/next
This is an implicit agreement that keeps all the logic in the User model creation.

**Comment by jum-s** [comment]
A quick look at what user stories look like : 

As a new user, 
when i click on Sign up
it creates a user store 
and ties it to the ID provider
and redirects to TOS
when i click on `Agree` TOS button
it redirect to the editor

As a new user, 
when i click on **disagree** TOS button
it deletes the user, 
and detroys the session 
and redirects to `/`

As an already authenticated user,
when i click on the editor button,
it automatically sign in and redirect to the editor

As an already authenticated user,
when i click on the logout button,
it destroys the session and redirect me to `/`

As a logged in user that wants to login with another identity,
when i go the ID provider setting 
and untie/remove ID provider link to data.transformap.co
when I click on the sign in button
it redirects to provider authentcation page

**Comment by acorbi** [comment]
@jum-s 

> A quick look at what user stories look like :

Can you please link me to the place where can I find these?

> I do not quite understand the necessity of this agreedTos storage while it could be implicit, with a 'less data storing' solution.
> When a new user signs up, they are sent to the provider page agreement
> Then, if a new user is created, they are redirected to TOS with an 'Agree' button.

The alternative you are proposing here sounds good to me. Where I cannot connect the dots though is on the *user creation / signup* part. https://tree.taiga.io/project/transformap/task/415 and https://tree.taiga.io/project/transformap/task/427 only mention Authentication (login) not user creation. So....

**Where are users being created?** I haven't thought much about this yet, but let me think aloud here... Since we currently use gitlab as "authentication method", users would need to have an account there (in our case, the provider is lab.allmende.io). The missing piece (for me, at this point) is then a "registration/signup" UI which lets the user register on lab.allmende.io and (as you propose) accept the TOS. Then they would be properly registered users and authorized to make changes and the only thing they need to do in the editor is login, which we already have.

**Summary**: I'm happy with scratching the `agreedTos` and `showTosMessage` parts of my implementation but still I'm unclear about user creation / signup.

maybe something @almereyda can shed some light on?

**Comment by jum-s** [comment]
@acorbi 
User already has a table which stores provider information in order to have a session.
The user stories are here to give idea of how it should work seamlessly. It's not integrated yet (you can check that [`/auth`](https://transformap-data.apps.allmende.io/secretPage) path already redirect to provide rinformation. That's the only feature fully implemented so far because completly independant from the editor. Then the login button redirect to this URL.

**Comment by acorbi** [comment]
@jum-s 

> So far, User record are immutable because it only stores ID provider references.
> User already has a table which stores provider information in order to have a session.

Understood, BUT how do the records on this `user` table get created? Am I overseeing something or there is currently no concept/plan for user sign-up? 

In my understanding we just have a concept/plan for the login, not for user creation. So, how about the following:

- I put 1-2 hours coding and add a sign-up button to the editor (close to the login button) which then lets users create an account after agreeing the TOS (which could then be embedded in a scrollable textarea) specifying username and email and POSTing them to `user` endpoint.

- I remove the `agreedTos` and `showTosMessage` parts of my implementation 

@almereyda what do you think?

**Comment by jum-s** [comment]
It was reported multiple times including [here](https://hack.allmende.io/transformaps-20170914-development#users-records) & [here](https://tree.taiga.io/project/transformap/task/416) that a user store get created when login for first time. It is store on a `User` database which stores an object that looks like this [fixture](https://github.com/jum-s/data.transformap.co/blob/master/test/fixtures/user-new-to-create-for-api.json).

To me, a resource server needs to match Id provider data against its own stored data, which garantees the 'rightfulness' of the identity to the editor and maintain a session from pages to pages. Passport integrates all of that seamlessly within one library.

**Pull request merged**